Introduce XPU implementation for PyTorch ATen operators

fengyuan14 opened the pull request:
As a follow-up to #114835 and #119682, we add limited ATen operators implementation for XPU. With this PR, the blocking issue for oneDNN operations and Inductor XPU backend will be resolved as the two components depend on these operations to support its basic features, respectively.

The added ATen operators include:

- `copy_`, `_to_copy`, `_copy_from_and_resize`, , `clone`
- `view`, `view_as_real`, `view_as_complex`,
- `as_strided`, `_reshape_alias`, `resize_`, `resize_as_`,
- `add`/`add_`, `sub`/`sub_`, `mul`/`mul_`, `div`/`div_`, `abs`, 
- `empty`, `empty_strided`, 
- `fill_`, `zeros_`.

EikanWang reviewed (changes requested):
Please add test cases for XPU ops.

fengyuan14 committed:
Update caffe2/CMakeLists.txt

Co-authored-by: Eikan Wang <eikan.wang@intel.com>

fengyuan14 committed:
Update caffe2/CMakeLists.txt

Co-authored-by: Eikan Wang <eikan.wang@intel.com>

fengyuan14 committed:
Update caffe2/CMakeLists.txt

Co-authored-by: Eikan Wang <eikan.wang@intel.com>

fengyuan14 committed:
Introduce XPU implementation for PyTorch ATen operators

A submodule is introduced as XPU implementation of PyTorch ATen, `third_party/torch-xpu-ops`.
So far, enabled limited ATen operators, including tensor factory operators, copy, basic binary operators and fill.

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Introduce XPU implementation for PyTorch ATen operators

A submodule is introduced as XPU implementation of PyTorch ATen, `third_party/torch-xpu-ops`.
So far, enabled limited ATen operators, including tensor factory operators, copy, basic binary operators and fill.

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Revert "Update caffe2/CMakeLists.txt"

This reverts commit 6a4ee5e2cfe593d71b3b9840048cfb66d80c76f3.

fengyuan14 committed:
Revert "Update caffe2/CMakeLists.txt"

This reverts commit 6a4ee5e2cfe593d71b3b9840048cfb66d80c76f3.

fengyuan14 committed:
add comments

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
rebase torch-xpu-ops

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
rebase torch-xpu-ops

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
rebase torch-xpu-ops for SYCL source incremental compilation

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
rebase torch-xpu-ops for SYCL source incremental compilation

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Bypass WERROR in ci for XPU build

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
rebase torch-xpu-ops for new supported operators

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
rebase torch-xpu-ops for new supported operators

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Rebase torch-xpu-ops for new supported ops and add simple unit tests

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Rebase torch-xpu-ops for new supported ops and add simple unit tests

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Correct variable definition in unit test

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Fix lint error

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

malfet commented on .gitmodules:
```diff
@@ -149,3 +149,6 @@
 [submodule "third_party/mimalloc"]
 	path = third_party/mimalloc
 	url = https://github.com/microsoft/mimalloc.git
+[submodule "third_party/torch-xpu-ops"]
+	path = third_party/torch-xpu-ops
+	url = https://github.com/intel/torch-xpu-ops.git
```
No, please let's not do that, but rather figure out some sort of a soft-dependency mechanism during the checkout

EikanWang replied:
@fengyuan14 , let's try `add_custom_target`

malfet reviewed (changes requested):
Let's not add additional submodules, but figure out some sort of a soft checkout mechanism

Also, this repository feels like a huge copy-n-paste of an already existing code (i.e. `src/aten/sycl/Loops.h` looks very similar to `aten/src/ATen/cuda/Loops.cuh`, which makes me wonder if some of this code can be better reused between the two somehow)

fengyuan14 commented:
> Also, this repository feels like a huge copy-n-paste of an already existing code (i.e. `src/aten/sycl/Loops.h` looks very similar to `aten/src/ATen/cuda/Loops.cuh`, which makes me wonder if some of this code can be better reused between the two somehow)

Hi, @malfet ,
Regarding performance, even we have same concurrency algorithm, kernel codes could not be totally same. We need some HW specific adjustments. Just like `aten/src/ATen/cuda/CUDALoops.cuh` and `aten/src/ATen/native/cuda/ROCmLoops.cuh`.

The current `Loops` implementation in `torch-xpu-ops` is similar as `aten/src/ATen/cuda/Loops.cuh`. BTW, we are porting Intel GPU specific optimized code-paths gradually from `intel-extension-for-pytorch` to align performance (`torch-xpu-ops` vs `intel-extension-for-pytorch`)

For long term, regarding in-tree `xpu`, we could abstract device agnostic parts for kernels.

Thanks.

EikanWang commented:
> Let's not add additional submodules, but figure out some sort of a soft checkout mechanism

@malfet , thanks for your comments. May I know your concerns about the git submodule integration mechanism? Regarding the soft checkout mechanism, we might consider using CMake's `ExternalProject` as a solution. This would allow us to clone XPU Aten operations during the build process if `USE_XPU` is set to `ON`. Apart from the ExternalProject approach, do you have any other suggestions or proposals?

albanD commented on test/test_xpu.py:
```diff
@@ -126,6 +129,170 @@ def test_generator(self):
         self.assertEqual(2024, torch.xpu.initial_seed())
 
+    # Simple cases for operators
+    # TODO: Reusing PyTorch test_ops.py to improve coverage
```
Please use https://github.com/pytorch/pytorch/blob/85dc2543640e62f416549658fb011c367fcb06b7/test/test_ops.py#L281-L310 and replace cuda with xpu.
You can add a local not-implemented and xfail list for everything that is not yet implemented / expected to fail for now.

fengyuan14 replied:
We plan to migrate to `test_ops.py` when we have better operator coverage.

In the first PR, we only have a few operators supported in `torch-xpu-ops`. The not-implemented list is too long and too mutable to maintain. So we follow `pytorch/test/test_mps.py` in the PR, providing xpu specific operator cases.

EikanWang replied:
@albanD , I just submitted a PR https://github.com/pytorch/pytorch/pull/121878 to extend the testing infrastructure to support XPU. From the execution perspective, I'm proposing a staging plan here. I would appreciate any comments.

- Stage 0 - Add the `test_xpu.py` dedicated to XPU to validate limited aten operations for XPU and XPU-specific cases
- Stage 1 - When the `ops` coverage for xpu is >= 50%, we migrate the test cases from `test_xpu.py` by utilizing `test_ops.py` w/ xpu-specific op list
- Stage 2 - Fully utilize `test_ops.py` and `test_xpu.py` only keeps the XPU-specific cases

In parallel, we will continuously refactor the test infrastructure aiming to support different devices in general.

@albanD , any comments?

albanD replied:
> The not-implemented list is too long and too mutable to maintain.

Just go with an allow-list that should be just "add" right now I guess.

> So we follow pytorch/test/test_mps.py

We should not. The takeaway from that approach is that it is not working well, we are still finding a lot of silent correctness issues in basic op stemming from the fact that we didn't start with ops directly. And now we have a lot of ops that are partially broken in weird ways.

> I'm proposing a staging plan here

For the XPU/CPU matching test, I don't think we have to move it to test_ops.py tbh. It is fine to keep it in test_xpu.py.
We do want to add coverage there for the other tests for autograd, etc though indeed.

EikanWang replied:
Thanks for your comments @albanD . Let me confirm if my understanding is correct.

- Keep `test_xpu.py`
- Refine the test cases of `test_xpu.py` by reusing the logic of `test_ops.py` as much as possible
  - Replace the `@onlyCUDA` with `@onlyXPU`
  - Pass the op list to `@ops` for xpu

The partial detailed implementation in `test_xpu.py` could be as follows.
```python
any_common_cpu_xpu_one = OpDTypes.any_common_cpu_cuda_one
_xpu_op_list = ["add", "mul"]
_xpu_ops = [_op for _op in _ops_and_refs_with_no_numpy_ref if _op.name in _xpu_op_list]

@unMarkDynamoStrictTest
class TestXPUCommon(TestCase):
    exact_dtype = True

    # Tests that the cpu and gpu results are consistent 
    @onlyXPU
    @suppress_warnings
    @slowTest
    @ops(_xpu_ops , dtypes=any_common_cpu_xpu_one)
    def test_compare_cpu(self, device, dtype, op):

        def to_cpu(arg):
            if isinstance(arg, torch.Tensor):
                return arg.to(device='cpu')
            return arg

        samples = op.reference_inputs(device, dtype)

        for sample in samples:
            cpu_sample = sample.transform(to_cpu)
            xpu_results = op(sample.input, *sample.args, **sample.kwargs)
            cpu_results = op(cpu_sample.input, *cpu_sample.args, **cpu_sample.kwargs)

            xpu_results = sample.output_process_fn_grad(xpu_results)
            cpu_results = cpu_sample.output_process_fn_grad(cpu_results)

            # Lower tolerance because we are running this as a `@slowTest`
            # Don't want the periodic tests to fail frequently
            self.assertEqual(xpu_results, cpu_results, atol=1e-3, rtol=1e-3)
```


@albanD , do I capture your point?

albanD replied:
Yes looks good!

EikanWang replied:
Thanks for your confirm.

EikanWang commented on caffe2/CMakeLists.txt:
```diff
@@ -1090,6 +1090,28 @@ if(USE_XPU)
   torch_compile_options(torch_xpu)  # see cmake/public/utils.cmake
   target_compile_options_if_supported(torch_xpu "-Wno-deprecated-copy")  # see cmake/public/utils.cmake
   target_compile_definitions(torch_xpu PRIVATE USE_XPU)
+
```
```suggestion
  if(NOT EXISTS "${TORCH_ROOT}/third_party/torch-xpu-ops/.git")
      execute_process(
          COMMAND git clone https://github.com/intel/torch-xpu-ops "${TORCH_ROOT}/third_party/torch-xpu-ops"
      )
  else()
      execute_process(
          COMMAND git -C "${third_party}/torch-xpu-ops" pull
      )
  endif()
```

EikanWang replied:
In general, @malfet , are you okay with the suggested change?

EikanWang replied:
@malfet, we have applied the change first to follow the soft-check mechanism. We would appreciate it if you had any implementation proposals :)

malfet replied:
Yes, this sounds fine to me

EikanWang replied:
Thanks, we will update the PR and keep you posted.

EikanWang replied:
We are working on the code and data separation. However, we have yet to find a proper place to put the commit pin file.  It kind of conflicts with the current PyTorch CMake design if we put the version under the CMake folder directly.

There are two places in the stock PyTorch to host commit pin files for different packages: `.ci` and `.github`. `.ci` is dedicated to ci. Therefore, we add a xpu commit pin to the `.github` folder now.

And any suggestions are appreciated.

malfet replied:
You can put it in `third_party` folder, which feels like a natural place to keep both data and metadata

EikanWang replied:
Good idea! Thanks for the suggestion. We will put the `xpu.txt` under `third_party`

EikanWang replied:
@malfet , we have refined the PR. May I know if you have any other comments?

fengyuan14 commented on .ci/pytorch/build.sh:
```diff
@@ -254,7 +254,8 @@ else
     if [[ "$BUILD_ENVIRONMENT" != *rocm*  &&
-          "$BUILD_ENVIRONMENT" != *xla* ]]; then
+          "$BUILD_ENVIRONMENT" != *xla* &&
+          "$BUILD_ENVIRONMENT" != *xpu* ]]; then
```
Hi,
Compilation verification for SYCL toolkit setup fails (SYCLToolkit.cmake) when ERROR=1. Will fix.

No source code compilation failure.

fengyuan14 committed:
Use soft-dependency mechanism instead of PyTorch git submodule

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Use soft-dependency mechanism instead of PyTorch git submodule

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Improve XPU operators test with common operators test architecture

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Improve XPU operators test with common operators test architecture

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Enble device type test for TestXpu

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Enble device type test for TestXpu

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
add cases for XPU

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Fix lint error

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Fix lint error

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

EikanWang reviewed (commented):
Please update the test cases a little bit

fengyuan14 committed:
Remove test_dtype and will retrieve it once we have fancy test frame to reuse cases.

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Remove test_dtype and will retrieve it once we have fancy test frame to reuse cases.

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
reduce test scope of XPU device

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
reduce test scope of XPU device

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
remove cases bypassing XPU

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
fix non-imported module

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
Fix lint error, remove no needed python modules

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

fengyuan14 committed:
fix lint error

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

atalman commented:
@pytorchbot rebase

EikanWang committed:
Move xpu commit pin file from .github to third_party

EikanWang committed:
rebase to main

albanD reviewed (commented):
The testing piece looks good to me. The next step there is going to have xfail for everything not in this list to make sure we know when coverage is changing.

I'll let @malfet take a final look at the submodule work and accept the PR.

EikanWang commented:
Thanks, @albanD . We will submit another pr to follow up the next step by marking all the other operations that have not been supported for xpu as xfail.

fengyuan14 committed:
Pin torch-xpu-ops on a specified commit with separating CMake source and commit info

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

EikanWang commented:
@pytorchbot rebase

EikanWang commented:
@pytorchbot rebase

fengyuan14 commented:
@pytorchbot rebase

atalman reviewed (approved):
LGTM

EikanWang commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but I need to revert it to resolve a conflict in trunk https://github.com/pytorch/pytorch/pull/121794#issuecomment-2013434523.  Please help reland the change after' -c weird

EikanWang commented:
@pytorchbot merge

EikanWang commented:
@pytorchbot merge -i
